File: retail/reaction.go
package retail

import "math"

// Signal bundles the inputs an archetype's reaction function can use.
// Not every archetype uses every field — e.g. Degenerate only cares
// about Volatility, MomentumChaser/Contrarian/PanicProne only care
// about Momentum.
type Signal struct {
	// Momentum is a simple rate-of-change measure (see
	// technicals.Momentum), positive for a rising market, negative for
	// falling.
	Momentum float64

	// Volatility is a normalized measure of how wide recent price
	// action has been (e.g. derived from Bollinger band width or ATR
	// relative to price) — what Degenerate reacts to regardless of
	// direction.
	Volatility float64

	// FundamentalGap is (ReportedValue-price)/price, the same shape of
	// signal HedgeFund uses, but read from ReportedValue rather than
	// TrueValue. Zero/unused for TechnicalOnly bots.
	FundamentalGap float64
}

// ActionProbabilities is a buy/sell/hold probability triple that sums
// to 1.0, used to weight a single random draw into a decision.
type ActionProbabilities struct {
	Buy  float64
	Sell float64
	Hold float64
}

// normalize rescales three non-negative weights so they sum to 1,
// guarding against a degenerate all-zero input by returning a pure
// hold.
func normalize(buy, sell, hold float64) ActionProbabilities {
	total := buy + sell + hold
	if total <= 0 {
		return ActionProbabilities{Hold: 1}
	}
	return ActionProbabilities{Buy: buy / total, Sell: sell / total, Hold: hold / total}
}

// sigmoid maps any real value into (0, 1), used to turn an unbounded
// signal strength into a bounded probability weight.
func sigmoid(x float64) float64 {
	return 1 / (1 + math.Exp(-x))
}

// ReactionFunc computes buy/sell/hold probabilities for a given
// archetype and signal. baseActivity scales overall willingness to
// act at all (vs. hold) — lower for Disciplined, higher for
// Degenerate/Beginner-flavored archetypes — kept as a parameter here
// rather than hardcoded per archetype, since SkillTier can also
// influence overall activity level independent of Archetype.
type ReactionFunc func(s Signal, baseActivity float64) ActionProbabilities

// reactionFuncs maps each Archetype to its ReactionFunc. Kept as a
// lookup table rather than a switch inside a single function so each
// archetype's logic is independently readable and testable.
var reactionFuncs = map[Archetype]ReactionFunc{
	MomentumChaser: func(s Signal, baseActivity float64) ActionProbabilities {
		// Buys strength, sells weakness: momentum directly drives the
		// buy/sell weight split, symmetric in both directions.
		strength := sigmoid(s.Momentum * 20) // scale momentum (a small fraction) into a meaningful sigmoid input
		buy := baseActivity * strength
		sell := baseActivity * (1 - strength)
		hold := 1 - baseActivity
		return normalize(buy, sell, hold)
	},

	Contrarian: func(s Signal, baseActivity float64) ActionProbabilities {
		// Mirror of MomentumChaser: buys weakness, sells strength.
		strength := sigmoid(s.Momentum * 20)
		buy := baseActivity * (1 - strength)
		sell := baseActivity * strength
		hold := 1 - baseActivity
		return normalize(buy, sell, hold)
	},

	PanicProne: func(s Signal, baseActivity float64) ActionProbabilities {
		// Asymmetric: small positive momentum registers mild buy response,
		// but negative momentum is amplified sharply into a sell response.
		if s.Momentum >= 0 {
			buy := baseActivity * (0.10 + 0.30*sigmoid(s.Momentum*10))
			sell := baseActivity * 0.05
			hold := 1 - buy - sell
			return normalize(buy, sell, hold)
		}
		amplified := math.Min(1, math.Abs(s.Momentum)*40) // sharp amplification of drops
		sell := baseActivity * (0.20 + 0.80*amplified)
		hold := 1 - sell
		return normalize(0, sell, hold)
	},

	Disciplined: func(s Signal, baseActivity float64) ActionProbabilities {
		// Small, dampened reactions across the board with steady baseline engagement.
		dampened := baseActivity * 0.35
		strength := sigmoid(s.Momentum * 20)
		buy := dampened * strength
		sell := dampened * (1 - strength)
		hold := 1 - dampened
		return normalize(buy, sell, hold)
	},

	Degenerate: func(s Signal, baseActivity float64) ActionProbabilities {
		// Reacts to volatility itself with a reliable baseline activity floor (0.35)
		// ensuring ongoing speculative retail flow so the market never permanently dies.
		volDriven := baseActivity * (0.35 + 0.65*math.Min(1, s.Volatility*8))
		directionLean := sigmoid(s.Momentum * 10) // mild lean, not the main driver
		buy := volDriven * directionLean
		sell := volDriven * (1 - directionLean)
		hold := 1 - volDriven
		return normalize(buy, sell, hold)
	},
}

// ReactionFor returns the ReactionFunc for a given archetype.
func ReactionFor(a Archetype) ReactionFunc {
	return reactionFuncs[a]
}

File: src/retail/reaction.rs
use super::taxonomy::Archetype;

#[derive(Debug, Clone, Copy, Default)]
pub struct Signal {
    pub momentum: f64,
    pub volatility: f64,
    pub fundamental_gap: f64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ActionProbabilities {
    pub buy: f64,
    pub sell: f64,
    pub hold: f64,
}

fn normalize(buy: f64, sell: f64, hold: f64) -> ActionProbabilities {
    let total = buy + sell + hold;
    if total <= 0.0 {
        ActionProbabilities {
            buy: 0.0,
            sell: 0.0,
            hold: 1.0,
        }
    } else {
        ActionProbabilities {
            buy: buy / total,
            sell: sell / total,
            hold: hold / total,
        }
    }
}

fn sigmoid(x: f64) -> f64 {
    1.0 / (1.0 + (-x).exp())
}

pub fn reaction_for(archetype: Archetype, s: Signal, base_activity: f64) -> ActionProbabilities {
    match archetype {
        Archetype::MomentumChaser => {
            let strength = sigmoid(s.momentum * 20.0);
            let buy = base_activity * strength;
            let sell = base_activity * (1.0 - strength);
            let hold = 1.0 - base_activity;
            normalize(buy, sell, hold)
        }
        Archetype::Contrarian => {
            let strength = sigmoid(s.momentum * 20.0);
            let buy = base_activity * (1.0 - strength);
            let sell = base_activity * strength;
            let hold = 1.0 - base_activity;
            normalize(buy, sell, hold)
        }
        Archetype::PanicProne => {
            if s.momentum >= 0.0 {
                let buy = base_activity * (0.10 + 0.30 * sigmoid(s.momentum * 10.0));
                let sell = base_activity * 0.05;
                let hold = 1.0 - buy - sell;
                normalize(buy, sell, hold)
            } else {
                let amplified = (s.momentum.abs() * 40.0).min(1.0);
                let sell = base_activity * (0.20 + 0.80 * amplified);
                let hold = 1.0 - sell;
                normalize(0.0, sell, hold)
            }
        }
        Archetype::Disciplined => {
            let dampened = base_activity * 0.35;
            let strength = sigmoid(s.momentum * 20.0);
            let buy = dampened * strength;
            let sell = dampened * (1.0 - strength);
            let hold = 1.0 - dampened;
            normalize(buy, sell, hold)
        }
        Archetype::Degenerate => {
            let vol_driven = base_activity * (0.35 + 0.65 * (s.volatility * 8.0).min(1.0));
            let direction_lean = sigmoid(s.momentum * 10.0);
            let buy = vol_driven * direction_lean;
            let sell = vol_driven * (1.0 - direction_lean);
            let hold = 1.0 - vol_driven;
            normalize(buy, sell, hold)
        }
    }
}
